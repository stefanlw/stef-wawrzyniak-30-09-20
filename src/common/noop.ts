// Do nothing
export const noop = () => {};
