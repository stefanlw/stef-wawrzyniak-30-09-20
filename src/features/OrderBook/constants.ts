import Config from 'react-native-config';

export const WEBSOCKET_ADDRESS: string = Config.WEBSOCKET_ADDRESS;

export const UPDATE_INTERVAL = 500;
