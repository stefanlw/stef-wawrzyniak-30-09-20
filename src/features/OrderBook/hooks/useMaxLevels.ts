import { useCallback, useState } from 'react';
import { LayoutChangeEvent } from 'react-native';
import { ORDER_BOOK_ROW_HEIGHT } from '~/components/OrderBook/Row/OrderBookRow';
import { Levels } from '../types/OrderBookTypes';

type UseMaxLevelsReturnType = {
  getMaxLevelsOnLayout: (event: LayoutChangeEvent) => void;
  trimToMaxLevels: (levels: Levels) => Levels;
};

export const useMaxLevels = (): UseMaxLevelsReturnType => {
  const [maxLevels, setMaxLevels] = useState<number>(0);

  const getMaxLevelsOnLayout = useCallback(({ nativeEvent }) => {
    const headerRows = ORDER_BOOK_ROW_HEIGHT * 2;
    setMaxLevels(
      Math.floor(
        (nativeEvent.layout.height - headerRows) / ORDER_BOOK_ROW_HEIGHT,
      ) / 2,
    );
  }, []);

  const trimToMaxLevels = useCallback(
    (levels: Levels) => {
      return levels.slice(0, maxLevels);
    },
    [maxLevels],
  );

  return {
    getMaxLevelsOnLayout,
    trimToMaxLevels,
  };
};
