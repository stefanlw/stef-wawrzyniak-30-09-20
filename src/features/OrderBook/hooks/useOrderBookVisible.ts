import { useEffect } from 'react';
import { AppState } from 'react-native';
import { useNavigation } from '@react-navigation/native';

export const useOrderBookVisible = (
  onFocus: () => void,
  onBlur: () => void,
) => {
  const navigation = useNavigation();
  useEffect(() => {
    const subscription = AppState.addEventListener('change', nextAppState => {
      if (nextAppState === 'active') {
        onFocus();
      } else {
        onBlur();
      }
    });

    return subscription.remove;
  }, [onFocus, onBlur]);

  useEffect(() => {
    // handle change between screens
    const unsubscribeFocus = navigation.addListener('focus', () => {
      onFocus();
    });

    const unsubscribeBlur = navigation.addListener('blur', () => {
      onBlur();
    });

    return () => {
      unsubscribeFocus();
      unsubscribeBlur();
    };
  }, [navigation, onFocus, onBlur]);
};
