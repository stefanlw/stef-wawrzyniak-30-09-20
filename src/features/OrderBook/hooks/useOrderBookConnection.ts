import { useCallback, useEffect, useRef, useState } from 'react';
import { WEBSOCKET_ADDRESS } from '../constants';
import {
  FeedDeltaMessage,
  FeedEvent,
  FeedSnapshot,
  Market,
} from '../types/OrderBookTypes';
import {
  createWebSocketConnection,
  isConnectionOpen,
  sendWSMessage,
} from './websocketUtils';

type UseOrderBookConnectionReturnType = {
  connect: () => void;
  disconnect: () => void;
  disconnected: boolean;
  error: string;
};

export const useOrderBookConnection = (
  market: Market,
  handleOrderBookUpdate: (data: FeedSnapshot | FeedDeltaMessage) => void,
  onNewSubscription: () => void,
): UseOrderBookConnectionReturnType => {
  const [disconnected, setDisconnected] = useState<boolean>(false);
  const [error, setError] = useState<string>('');

  const ws = useRef<WebSocket | null>(null);

  useEffect(() => {
    if (ws.current && isConnectionOpen(ws.current)) {
      handleContractChange();
    }
  }, [market]);

  useEffect(() => {
    connect();
    return disconnect;
  }, []);

  const connect = () => {
    if (ws.current && isConnectionOpen(ws.current)) return;
    setError('');
    setDisconnected(false);

    if (!WEBSOCKET_ADDRESS) {
      throw new Error('WEBSOCKET_ADDRESS environment variable is missing');
    }

    ws.current = createWebSocketConnection(WEBSOCKET_ADDRESS);

    ws.current.onopen = () => {
      sendWSMessage(ws.current, 'subscribe', [market]);
    };

    ws.current.onmessage = e => {
      try {
        const data = JSON.parse(e.data);
        if (data.event) {
          handleSubcriptionEvent(data);
        } else {
          handleOrderBookUpdate(data);
        }
      } catch (e) {
        setError(`Caught onMessage handler error: ${e.message}`);
      }
    };

    const handleSubcriptionEvent = (data: FeedEvent) => {
      if (data.event === 'subscribed') {
        onNewSubscription();
      } else if (data.event === 'unsubscribed') {
        onNewSubscription();
      }
    };

    ws.current.onerror = e => {
      console.log(`WebSocket error: ${e.message}`);
      setError(e.message);
      ws.current = null;
    };

    ws.current.onclose = e => {
      console.log(`WebSocket connection closed ${e.code}: ${e.reason}`);
      setDisconnected(true);
      ws.current = null;
    };
  };

  const disconnect = useCallback(() => {
    if (ws.current && isConnectionOpen(ws.current)) {
      ws.current.close();
    }
  }, [ws.current]);

  const handleContractChange = () => {
    sendWSMessage(ws.current, 'unsubscribe', [
      market === 'PI_XBTUSD' ? 'PI_ETHUSD' : 'PI_XBTUSD',
    ]);

    sendWSMessage(ws.current, 'subscribe', [market]);
  };

  return {
    connect,
    disconnect,
    disconnected,
    error,
  };
};
