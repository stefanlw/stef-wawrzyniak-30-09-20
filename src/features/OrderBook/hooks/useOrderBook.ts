import { useCallback, useEffect, useRef, useState } from 'react';
import { useOrderBookConnection } from './useOrderBookConnection';
import { buildOrderBook, mergeOrders } from '../utils/OrderBookUtils';
import { UPDATE_INTERVAL } from '../constants';
import {
  FeedDeltaMessage,
  FeedMessage,
  FeedSnapshot,
  Market,
  OrderBookState,
} from '../types/OrderBookTypes';

type UseOrderBookReturnType = {
  connect: () => void;
  disconnect: () => void;
  disconnected: boolean;
  toggleMarket: () => void;
  orderBook: OrderBookState;
  error: string;
};

export const DEFAULT_ORDER_BOOK_STATE = {
  asks: [],
  bids: [],
  spread: {
    value: '',
    percentage: '',
  },
};

export const useOrderBook = (): UseOrderBookReturnType => {
  const [market, setMarket] = useState<Market>('PI_XBTUSD');
  const [orderBook, setOrderBook] = useState<OrderBookState>(
    DEFAULT_ORDER_BOOK_STATE,
  );

  const bidsBuffer = useRef({});
  const asksBuffer = useRef({});
  const updateInterval = useRef<number>(0);

  useEffect(() => {
    updateInterval.current = setInterval(() => {
      updateOrderBook();
    }, UPDATE_INTERVAL);
    return () => {
      clearInterval(updateInterval.current);
    };
  }, []);

  const updateOrderBook = () => {
    const newOrderBook = buildOrderBook(asksBuffer.current, bidsBuffer.current);
    setOrderBook(newOrderBook);
  };

  const handleOrderBookUpdate = (
    data: FeedSnapshot | FeedDeltaMessage,
  ): void => {
    const { asks, bids } = data;
    bidsBuffer.current = mergeOrders(bids, bidsBuffer.current);
    asksBuffer.current = mergeOrders(asks, asksBuffer.current);

    if (isSnapshot(data)) {
      updateOrderBook();
    }
    // TODO: test performance on low end device
  };

  const resetOrderBookState = () => {
    bidsBuffer.current = {};
    asksBuffer.current = {};
    setOrderBook(DEFAULT_ORDER_BOOK_STATE);
  };

  const { connect, disconnect, disconnected, error } = useOrderBookConnection(
    market,
    handleOrderBookUpdate,
    resetOrderBookState,
  );

  useEffect(() => {
    if (error) {
      setOrderBook(DEFAULT_ORDER_BOOK_STATE);
    }
  }, [error]);

  const toggleMarket = useCallback(() => {
    if (market === 'PI_XBTUSD') {
      setMarket('PI_ETHUSD');
    } else if (market === 'PI_ETHUSD') {
      setMarket('PI_XBTUSD');
    } else {
      console.warn('Unhanled market state', market);
    }
  }, [market]);

  return {
    connect,
    disconnect,
    disconnected,
    toggleMarket,
    orderBook,
    error,
  };
};

const isSnapshot = (data: FeedMessage) => data.feed.includes('snapshot');
