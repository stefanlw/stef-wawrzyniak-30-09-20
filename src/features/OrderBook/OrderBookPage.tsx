import React, { useCallback } from 'react';
import { View, StyleSheet } from 'react-native';
import { colors } from '~/theme/colors';
import { spacing } from '~/theme/spacing';
import { useOrderBookVisible } from './hooks/useOrderBookVisible';
import { useOrderBook } from './hooks/useOrderBook';
import { Toast } from '~/components/Toast/Toast';
import { OrderBook } from '~/components/OrderBook/OrderBook';
import { PageHeader } from '~/components/Text/Text';
import { SafeAreaView } from 'react-native-safe-area-context';
import { Button } from '~/components/Button';
import { noop } from '~/common/noop';

export const OrderBookPage = () => {
  const { connect, disconnect, disconnected, orderBook, toggleMarket, error } =
    useOrderBook();

  useOrderBookVisible(noop, disconnect);

  return (
    <SafeAreaView style={styles.safeArea}>
      <View style={styles.container}>
        <View style={styles.headerContainer}>
          <PageHeader>Order Book</PageHeader>
        </View>
        <OrderBook orderBook={orderBook} />
      </View>
      <View style={styles.center}>
        <Button onPress={toggleMarket}>Toggle Feed</Button>
      </View>

      {/* TODO: improve error handling */}
      {(disconnected || !!error) && (
        <Toast
          message={`${
            disconnected ? 'Feed disconnected' : 'An error occurred'
          }, please press reconnect when ready`}
          ctaText="Reconnect"
          callback={connect}
        />
      )}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    backgroundColor: colors.darkBlue,
  },
  container: {
    flex: 1,
  },
  headerContainer: {
    marginHorizontal: spacing.xlarge,
    marginBottom: spacing.medium,
  },
  center: {
    justifyContent: 'center',
    alignItems: 'center',
  },
});
