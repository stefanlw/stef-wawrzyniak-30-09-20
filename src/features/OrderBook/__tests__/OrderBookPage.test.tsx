import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { MOCK_ORDER_BOOK_STATE } from '../../../../__mocks__/OrderBook';
import { OrderBookPage } from '../OrderBookPage';
import { useOrderBookVisible } from '../hooks/useOrderBookVisible';
import { useOrderBook } from '../hooks/useOrderBook';
import { noop } from '~/common/noop';

jest.mock('@react-navigation/native');

const mockConnectFn = jest.fn();
const mockDisconnectFn = jest.fn();

// TODO: extract + make utils of the following
jest.mock('../hooks/useOrderBook', () => ({
  ...jest.requireActual('../hooks/useOrderBook'),
  useOrderBook: jest.fn(() => ({
    connect: mockConnectFn,
    disconnect: mockDisconnectFn,
    disconnected: true,
    toggleMarket: jest.fn(),
    orderBook: MOCK_ORDER_BOOK_STATE,
  })),
}));

jest.mock('../hooks/useOrderBookConnection', () => ({
  ...jest.requireActual('../hooks/useOrderBookConnection'),
  useOrderBookConnection: jest.fn(() => ({
    connect: mockConnectFn,
    disconnect: jest.fn(),
    disconnected: true,
    error: jest.fn(),
  })),
}));

jest.mock('../hooks/useOrderBookVisible', () => ({
  ...jest.requireActual('../hooks/useOrderBookVisible'),
  useOrderBookVisible: jest.fn(),
}));

jest.mock('../hooks/useMaxLevels', () => ({
  ...jest.requireActual('../hooks/useMaxLevels'),
  useMaxLevels: jest.fn(() => ({
    trimToMaxLevels: jest.fn(levels => levels.slice(0, 5)),
  })),
}));

describe('OrderBookPage', () => {
  it('should render correctly', () => {
    const component = render(<OrderBookPage />);
    expect(component).toMatchSnapshot();
  });

  it('should call the useOrderBook hook', () => {
    render(<OrderBookPage />);
    expect(useOrderBook).toHaveBeenCalled();
  });

  it('should call the useOrderBookVisible hook', () => {
    render(<OrderBookPage />);
    expect(useOrderBookVisible).toHaveBeenCalledWith(noop, mockDisconnectFn);
  });

  describe('disconnected toast', () => {
    it('should display disconnected toast', () => {
      const component = render(<OrderBookPage />);
      const toast = component.getByText(
        'Feed disconnected, please press reconnect when ready',
      );
      expect(toast).toBeTruthy();
    });

    it('should call connect on reconnect press', () => {
      const component = render(<OrderBookPage />);
      const reconnectBtn = component.getByText('Reconnect');
      fireEvent.press(reconnectBtn);
      expect(mockConnectFn).toHaveBeenCalledTimes(1);
    });
  });
});
