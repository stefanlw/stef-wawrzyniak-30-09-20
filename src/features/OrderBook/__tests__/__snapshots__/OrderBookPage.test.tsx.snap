// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderBookPage should render correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#10141e",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 8,
          "marginHorizontal": 16,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 26,
            },
          ]
        }
      >
        Order Book
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            Object {
              "borderBottomColor": "#404854",
              "borderBottomWidth": 0.5,
              "borderTopColor": "#404854",
              "borderTopWidth": 0.5,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#404854",
                },
              ]
            }
          >
            PRICE
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#404854",
                },
              ]
            }
          >
            SIZE
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#404854",
                },
              ]
            }
          >
            TOTAL
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#351a23",
                "width": 399.6150907915614,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#b13135",
                },
                undefined,
              ]
            }
          >
            47,974.50
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            7.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            9,067.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#351a23",
                "width": 399.306575777164,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#b13135",
                },
                undefined,
              ]
            }
          >
            47,972.50
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            2,500.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            9,060.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#351a23",
                "width": 289.12264206381855,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#b13135",
                },
                undefined,
              ]
            }
          >
            47,970.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            112.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            6,560.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#351a23",
                "width": 284.18640183346065,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#b13135",
                },
                undefined,
              ]
            }
          >
            47,969.50
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            6,338.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            6,448.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#351a23",
                "width": 4.848093083387201,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#b13135",
                },
                undefined,
              ]
            }
          >
            47,969.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            110.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            110.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              Object {
                "alignItems": "center",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#404854",
                },
              ]
            }
          >
            Spread: 
            6
             (
            0.00
            %)
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#102e2d",
                "width": 4.848093083387201,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#12c48c",
                },
                undefined,
              ]
            }
          >
            47,963.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            110.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            110.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#102e2d",
                "width": 48.92166656872539,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#12c48c",
                },
                undefined,
              ]
            }
          >
            47,962.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            1,000.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            1,110.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#102e2d",
                "width": 342.9364752894165,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#12c48c",
                },
                undefined,
              ]
            }
          >
            47,961.50
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            6,671.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            7,781.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#102e2d",
                "width": 667.2298289945348,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#12c48c",
                },
                undefined,
              ]
            }
          >
            47,960.50
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            7,358.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            15,139.00
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 25,
              "paddingHorizontal": 40,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 25,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "#102e2d",
                "width": 750,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#12c48c",
                },
                undefined,
              ]
            }
          >
            47,956.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            1,878.00
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 3,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontVariant": Array [
                    "tabular-nums",
                  ],
                },
                Object {
                  "color": "#ffffff",
                },
                undefined,
              ]
            }
          >
            17,017.00
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 6,
              "justifyContent": "center",
              "padding": 12,
              "paddingHorizontal": 16,
            },
            Object {
              "backgroundColor": "#4b3cbb",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 14,
              "fontWeight": "600",
            }
          }
        >
          Toggle Feed
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "bottom": 50,
          "height": 50,
          "paddingHorizontal": 16,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 6,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "padding": 8,
          },
          Object {
            "width": 718,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 3,
          }
        }
      >
        <Text
          style={Object {}}
        >
          Feed disconnected, please press reconnect when ready
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <Text
          style={Object {}}
        >
          Reconnect
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
